cpp/src/arrow/csv/writer.cc

```diff
@@ -108,6 +108,17 @@ int64_t CountQuotes(std::string_view s) {
 constexpr int64_t kQuoteCount = 2;
 constexpr int64_t kQuoteDelimiterCount = kQuoteCount + /*end_char*/ 1;
 
+#ifndef CSV_WRITER_DISPATCH_STRING_ARRAY_TYPE
+#define CSV_WRITER_DISPATCH_STRING_ARRAY_TYPE(array, func, ...) \
```

Comment:
I'm afraid it would be a bit over-design. There's actually two method `PopulateRows` and `UpdateRowLengths` that would require such a dispatch. That means, I would need to define four visitor classes for two different methods in two different classes.

Any ideas? @bkietz

Reply:
Kindly ping @bkietz, do you think if we would just keep current implementation, for simplicity?

Thanks!

Reply:
Well, a visitor might be overkill, but it would be better to remove the macro and write the code explicitly, IMHO.

Reply:
I also thinks a template helper function or just write the code is a better way

Reply:
Fixed. @pitrou could please help to take another look? thanks!